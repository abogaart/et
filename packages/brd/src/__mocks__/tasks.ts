import { Tasks } from '@bloomreach/cli-tasks';

export function mockChoices(index: number): void {
  Tasks.choice = jest.fn().mockImplementation((
    _message: string,
    choices: { name: string; value: any }[],
    next: (choice: any) => Promise<void>,
  ) => next(choices[index].value));
}

export function mockCustomPath(...ids: string[]): void {
  Tasks.customPath = jest.fn().mockImplementation((
    confirmMsg: string,
    _pathLabel: string,
    next: (path: string) => Promise<void>,
  ) => {
    const pathId = ids.find((id: string) => confirmMsg.includes(id));
    return pathId
      ? next(`/custom-path/${pathId}`)
      : next('/custom-path');
  });
}
