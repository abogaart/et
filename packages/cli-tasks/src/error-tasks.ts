import { ListrTaskWrapper } from 'listr';

export function taskError(msg: string, task: ListrTaskWrapper, cmd: any): void {
  task.title = msg;
  cmd.exit(1);
}
