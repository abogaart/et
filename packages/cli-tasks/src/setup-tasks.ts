const listrInquirer = require('listr-inquirer');
const path = require('path');

export const Tasks = {
  choice(message: string, choices: string[], next: (path: string) => Promise<void>) {
    return listrInquirer([
      {
        type: 'list',
        name: 'choice',
        message,
        choices,
      },
    ], async (answers: any) => next(answers.choice));
  },

  confirm(message: string) {
    return {
      type: 'confirm',
      name: 'continue',
      message,
    };
  },

  async customPath(
    confirmMsg: string,
    pathLabel: string,
    next: (path: string) => Promise<void>,
    error: () => Promise<void>) {
    return listrInquirer([
      Tasks.confirm(confirmMsg),
      {
        when: (answers: any) => answers.continue,
        type: 'input',
        name: 'customPath',
        message: pathLabel,
        validate: (str: string) => {
          if (str && str.trim() !== '') {
            return path.isAbsolute(str) || 'Path should be absolute';
          }
          return 'Value is required';
        },
      },
    ], async (answers: any) => (answers.continue === false
      ? error()
      : next(answers.customPath.trim())));
  },
};
