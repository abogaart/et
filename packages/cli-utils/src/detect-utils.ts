import * as findJavaHome from 'find-java-home';
import * as path from 'path';
import * as which from 'which';

export async function tryWhich(exec: string): Promise<string | null> {
  try {
    return await which(exec);
  } catch (ignore) {
    return null;
  }
}

export async function detectJavaPath(): Promise<string | null> {
  return new Promise((resolve) => {
    findJavaHome({ allowJre: false }, (err, home) => {
      if (err) {
        resolve(null);
      }
      resolve(path.resolve(home, 'bin', 'java'));
    });
  });
}
