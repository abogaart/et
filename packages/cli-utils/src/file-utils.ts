import * as fs from 'fs-extra';

export async function checkPathExists(filePath: string): Promise<boolean> {
  return fs.pathExists(filePath);
  // return new Promise((resolve: any) => {
  //   fs.access(filePath, fs.constants.F_OK, (error: Error | null) => {
  //     resolve(!error);
  //   });
  // });
}

export function findLinkedFile(file: string): string {
  if (!fs.lstatSync(file).isSymbolicLink()) return file;
  return findLinkedFile(fs.realpathSync(file));
}

export async function writeFile(path: string, data: string): Promise<boolean> {
  try {
    await fs.outputFile(path, data);
    return true;
  } catch (ignore) {
    // TODO: log error
    return false;
  }
}

export async function readFile(path: string): Promise<string | null> {
  try {
    const buffer = await fs.readFile(path);
    return buffer.toString('utf8');
  } catch (ignore) {
    // TODO: log error
    return null;
  }
}

export async function readJson(path: string): Promise<any | null> {
  try {
    const result = await fs.readJSON(path);
    return result;
  } catch (ignore) {
    // TODO: log error
    return null;
  }
}

export async function writeJson(path: string, data: any): Promise<boolean> {
  try {
    await fs.outputJSON(path, data);
    return true;
  } catch (ignore) {
    // TODO: log error
    return false;
  }
}
