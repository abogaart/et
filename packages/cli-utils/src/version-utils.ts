import * as execa from 'execa';

export async function getJavaVersion(path: string): Promise<string | null> {
  try {
    const { stdout, stderr } = await execa.sync(path, ['-version']);
    const out = stdout || stderr;
    if (out.startsWith('java version')) {
      return out.split('\n')[0].split(' ')[2].replace(/"/g, '');
    }
  // eslint-disable-next-line no-empty
  } catch (e) {}

  return null;
}

export async function getGitVersion(path: string): Promise<string | null> {
  try {
    const { stdout } = await execa.sync(path, ['--version']);
    if (stdout.startsWith('git version')) {
      return stdout.split(' ')[2];
    }
  // eslint-disable-next-line no-empty
  } catch (e) {}

  return null;
}

export async function getMavenVersion(path: string): Promise<string | null> {
  try {
    const { stdout, stderr } = await execa.sync(path, ['--version']);
    const out = stdout || stderr;
    if (out.startsWith('Apache Maven')) {
      return out.split('\n')[0].split(' ')[2];
    }
  // eslint-disable-next-line no-empty
  } catch (e) {}

  return null;
}
