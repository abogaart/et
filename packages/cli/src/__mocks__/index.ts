import * as path from 'path';

import { EtCommand, EtFlags, EtContext } from '../base';

export class TestCommand<Flags extends EtFlags> extends EtCommand<Flags> {
  static flags = EtCommand.flags;

  run = jest.fn();

  get context(): EtContext<Flags> {
    return this.getContext();
  }
}

export function fixturesPath(...paths): string {
  return path.resolve(__dirname, '../__fixtures__', ...paths);
}
