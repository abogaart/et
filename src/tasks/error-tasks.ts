import { ListrTaskWrapper } from 'listr';

export function taskError(msg: string, task: ListrTaskWrapper, cmd: any) {
  task.title = msg;
  cmd.exit(1);
}
