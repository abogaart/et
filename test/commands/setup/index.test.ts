import { test } from '@oclif/test';
const normalizeNewline = require('normalize-newline');

import { taskError } from '../../../src/tasks/error-tasks';
import { Tasks } from '../../../src/tasks/setup-tasks';
import { detectJavaPath, tryWhich } from '../../../src/utils/detect-utils';
import { checkPathExists, findLinkedFile } from '../../../src/utils/file-utils';
import { getGitVersion, getJavaVersion, getMavenVersion } from '../../../src/utils/version-utils';
import { jestExpect, mockListerLog } from '../../__mocks__';

mockListerLog();
jest.mock('listr-inquirer');
jest.mock('listr-verbose-renderer/lib/utils');
jest.mock('../../../src/utils/detect-utils');
jest.mock('../../../src/utils/file-utils');
jest.mock('../../../src/utils/version-utils');
jest.mock('../../../src/tasks/setup-tasks');
jest.mock('../../../src/tasks/error-tasks');

const actualCheckPathExists = jest.requireActual('../../../src/utils/file-utils').checkPathExists;
checkPathExists.mockImplementation((path: string) => actualCheckPathExists(path));

function mockChoices(index: number) {
  Tasks.choice.mockImplementation((_message: string, choices: string[], next: (path: string) => Promise<void>) => next(choices[index]));
}

function mockTryWhich(...includes: string[]) {
  tryWhich.mockImplementation(async (name: string) => includes.includes(name) ? `${name}.which` : null);
}

function mockCustomPath(...ids: string[]) {
  Tasks.customPath.mockImplementation((confirmMsg: string,
                                       _pathLabel: string,
                                       next: (path: string) => Promise<void>) => {
    for (const id of ids) {
      if (confirmMsg.includes(id)) {
        return next(`/custom-path/${id}`);
      }
    }
    return next('/custom-path');
  });
}

function mockJavaPath(value: string | null) {
  detectJavaPath.mockImplementation(async () => value);
}

function expectTaskError() {
  jestExpect(taskError.mock.calls.length).toBe(1);
  jestExpect(taskError.mock.calls[0][0]).toMatchSnapshot();
}

describe('setup -', () => {
  beforeEach(() => {
    taskError.mockClear();

    mockChoices(0);
    mockCustomPath('git', 'java', 'maven');
    mockTryWhich('git', 'java', 'mvn');
    mockJavaPath('./java-home');

    findLinkedFile.mockImplementation((name: string) => `${name}.canonical`);
    getGitVersion.mockReturnValue('1.2.3');
    getJavaVersion.mockReturnValue('2.3.4');
    getMavenVersion.mockReturnValue('3.4.5');
  });

  describe('initial setup -', () => {
    test
      .stdout()
      .command(['setup'])
      .it('finds global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(normalizeNewline(ctx.stdout)).toMatchSnapshot();
      });

    test
      .do(() => mockChoices(1))
      .stdout()
      .command(['setup'])
      .it('can select absolute path from global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(normalizeNewline(ctx.stdout)).toMatchSnapshot();
      });

    test
      .do(() => mockChoices(2))
      .stdout()
      .command(['setup'])
      .it('can select symlink resolved absolute path from global installed git, java and mvn executables', (ctx: any) => {
        jestExpect(normalizeNewline(ctx.stdout)).toMatchSnapshot();
      });

    test
      .do(() => mockTryWhich('git', 'mvn'))
      .stdout()
      .command(['setup'])
      .it('tries to find java-home when "which" command fails', (ctx: any) => {
        jestExpect(normalizeNewline(ctx.stdout)).toMatchSnapshot();
      });

    test
      .do(() => {
        mockTryWhich();
        mockJavaPath(null);
      })
      .stdout()
      .command(['setup'])
      .it('present option to set custom paths', (ctx: any) => {
        jestExpect(normalizeNewline(ctx.stdout)).toMatchSnapshot();
      });

    test
      .do(() => getGitVersion.mockReturnValue(null))
      .stdout()
      .stderr()
      .command(['setup'])
      .it('prints an error if git path is invalid', () => expectTaskError());

    test
      .do(() => getJavaVersion.mockReturnValue(null))
      .stdout()
      .stderr()
      .command(['setup'])
      .it('prints an error if java path is invalid', () => expectTaskError());

    test
      .do(() => getMavenVersion.mockReturnValue(null))
      .stdout()
      .stderr()
      .command(['setup'])
      .it('prints an error if maven path is invalid', () => expectTaskError());

    describe('custom path not set -', () => {
      beforeEach(() => {
        mockJavaPath(null);
        Tasks.customPath.mockImplementation((_confirmMsg: string,
                                             _pathLabel: string,
                                             _next: (path: string) => Promise<void>,
                                             error: () => Promise<void>) => {
          return error();
        });
      });

      test
        .do(() => mockTryWhich())
        .stdout()
        .stderr()
        .command(['setup'])
        .it('prints an error if custom git path is not set', () => expectTaskError());

      test
        .do(() => mockTryWhich('git'))
        .stdout()
        .stderr()
        .command(['setup'])
        .it('prints an error if custom java path is not set', () => expectTaskError());

      test
        .do(() => mockTryWhich('git', 'java'))
        .stdout()
        .stderr()
        .command(['setup'])
        .it('prints an error if custom mvn path is not set', () => expectTaskError());

    });
  });
});

describe('already installed', () => {
  test
    .stdout()
    .command(['setup', '--configFile', 'test/__fixtures__/et-installed'])
    .it('print merged config if already stalled', (ctx: any) => {
      jestExpect(ctx.stdout).toMatchSnapshot();
    });
});
