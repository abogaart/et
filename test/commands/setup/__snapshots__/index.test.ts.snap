// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`already installed print merged config if already stalled 1`] = `
"ET is installed with the following properties
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"bet-installed\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;

exports[`setup - initial setup - can select absolute path from global installed git, java and mvn executables 1`] = `
"
ET has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
Git found at 'git.which' (v1.2.3) [title changed]
Git found at 'git.which' (v1.2.3) [completed]
Detect Java executable [started]
Java detected at 'java.which' (v2.3.4) [title changed]
Java detected at 'java.which' (v2.3.4) [completed]
Detect Maven [started]
Maven detected at 'mvn.which' (v3.4.5) [title changed]
Maven detected at 'mvn.which' (v3.4.5) [completed]

Setup has successfully finished
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git.which\\"
  },
  \\"java\\": {
    \\"path\\": \\"java.which\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn.which\\"
  }
}
"
`;

exports[`setup - initial setup - can select symlink resolved absolute path from global installed git, java and mvn executables 1`] = `
"
ET has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
Git found at 'git.which.canonical' (v1.2.3) [title changed]
Git found at 'git.which.canonical' (v1.2.3) [completed]
Detect Java executable [started]
Java detected at 'java.which.canonical' (v2.3.4) [title changed]
Java detected at 'java.which.canonical' (v2.3.4) [completed]
Detect Maven [started]
Maven detected at 'mvn.which.canonical' (v3.4.5) [title changed]
Maven detected at 'mvn.which.canonical' (v3.4.5) [completed]

Setup has successfully finished
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git.which.canonical\\"
  },
  \\"java\\": {
    \\"path\\": \\"java.which.canonical\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn.which.canonical\\"
  }
}
"
`;

exports[`setup - initial setup - custom path not set - prints an error if custom git path is not set 1`] = `"Setup failed, please install Git first. For more info see https://www.atlassian.com/git/tutorials/install-git"`;

exports[`setup - initial setup - custom path not set - prints an error if custom java path is not set 1`] = `"Setup failed, please install Java first. For more info see https://www.java.com/en/download/help/download_options.xml"`;

exports[`setup - initial setup - custom path not set - prints an error if custom mvn path is not set 1`] = `"Setup failed, please install Maven first. For more info see https://maven.apache.org/install.html"`;

exports[`setup - initial setup - finds global installed git, java and mvn executables 1`] = `
"
ET has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
Git found at 'git' (v1.2.3) [title changed]
Git found at 'git' (v1.2.3) [completed]
Detect Java executable [started]
Java detected at 'java' (v2.3.4) [title changed]
Java detected at 'java' (v2.3.4) [completed]
Detect Maven [started]
Maven detected at 'mvn' (v3.4.5) [title changed]
Maven detected at 'mvn' (v3.4.5) [completed]

Setup has successfully finished
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;

exports[`setup - initial setup - present option to set custom paths 1`] = `
"
ET has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
Git found at '/custom-path/git' (v1.2.3) [title changed]
Git found at '/custom-path/git' (v1.2.3) [completed]
Detect Java executable [started]
Java detected at '/custom-path/java' (v2.3.4) [title changed]
Java detected at '/custom-path/java' (v2.3.4) [completed]
Detect Maven [started]
Maven detected at '/custom-path/maven' (v3.4.5) [title changed]
Maven detected at '/custom-path/maven' (v3.4.5) [completed]

Setup has successfully finished
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"/custom-path/git\\"
  },
  \\"java\\": {
    \\"path\\": \\"/custom-path/java\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"/custom-path/maven\\"
  }
}
"
`;

exports[`setup - initial setup - prints an error if git path is invalid 1`] = `"Failed to verify git executable at 'git'"`;

exports[`setup - initial setup - prints an error if java path is invalid 1`] = `"Failed to verify java executable at 'java'"`;

exports[`setup - initial setup - prints an error if maven path is invalid 1`] = `"Failed to verify Maven executable at 'mvn'"`;

exports[`setup - initial setup - tries to find java-home when "which" command fails 1`] = `
"
ET has not been installed yet, starting initial setup of version 0.0.0

Detect git executable [started]
Git found at 'git' (v1.2.3) [title changed]
Git found at 'git' (v1.2.3) [completed]
Detect Java executable [started]
Java detected at './java-home' (v2.3.4) [title changed]
Java detected at './java-home' (v2.3.4) [completed]
Detect Maven [started]
Maven detected at 'mvn' (v3.4.5) [title changed]
Maven detected at 'mvn' (v3.4.5) [completed]

Setup has successfully finished
{
  \\"env\\": \\"test\\",
  \\"installed\\": true,
  \\"version\\": \\"0.0.0\\",
  \\"git\\": {
    \\"path\\": \\"git\\"
  },
  \\"java\\": {
    \\"path\\": \\"./java-home\\"
  },
  \\"mvn\\": {
    \\"path\\": \\"mvn\\"
  }
}
"
`;
